enhancements/security/runtime-class-scc.md

```diff
@@ -0,0 +1,326 @@
+2. Specifying a set of items (`["foo"]`) requires pods in the SCC must request a runtime class that matches an item in the list, and cannot use the default.
+3. Specifying a set of items and an empty string (`["", "foo"]`) requires pods in the SCC must request a runtime class that matches an item in the list,
+or specify the default
+4. Specifying `["*"]` means pods in the SCC can request any runtime class, or omit requesting a custom one.
+Note, specifying `["*", "foo"]` is equivalent to specifying `["*"]`.
+
+The default value will be `[""]`, or a list with a single empty string, signifying users may not request any runtime class, and are required to use the default.
+
+This addition will break existing users of runtime classes. Luckily, Openshift does not yet ship any other runtime classes, other than the default.
+To be able to access different runtime classes, admins would have to add a custom MachineConfig to configure CRI-O to add a runtime class,
+and this customization is not supported.
+However, if admins happened to have added them, they will also have to allow their users to use them by specifying this field.
+
+#### Point Values:
+Openshift SCCs need to be ordered by relative point value of "how restrictive" they are (a system described [here][byrestrictions].
+This section describes how points will be assigned for `RequiredRuntimeClasses`.
+In general, a smaller set of RequiredRuntimeClasses is more restricted. For instance, a list that is a strict subset of another is more restrictive, and thus should have
+a lower score. A set that specifies `"*"` will have the maximum score added for `RequiredRuntimeClasses`.
+
+Taking the [current examples](#some-examples), there is no simple algorithm for how each runtime class is weighted against the others.
+For instance, an order could be argued (from least restrictive to most):
+`openshift-low-latency`, `"", `openshift-builder`, `openshift-sandboxed-containers`.
+- `openshift-low-latency` gives access to configuring the host in particular ways, and is definitely less restrictive than the other.
+- `openshift-builder` is certainly more restrictive than the default, because when `openshift-builder` is paired with any added capabilities, and the ability to run as root,
+it is more secure than the default paired with capabilities + run as root, as it is more secure to give a process capabilities/UID 0 in a user namespace, rather than
+the user namespace of the host.
+- `openshift-sandboxed-containers` are certainly the most restrictive, as it runs the process in an kernel separated container, protecting the host from a class of kernel
+vulnerabilities the default runtime is vulnerable to.
+
+Thus, a strategy similar to how [RunAsUser][run-as-user] could be implemented, where every new possible entry in AllowedRuntimeClasses must be given a point value.
```

Comment:
I don't think we should or even can assign values to known runtime classes. Admission should not know about runtime classes, i.e. the meaning of the strings, at all. Think about version skew cases between apiserver and kubelet and runtime.

Also I don't think the actual point values of individual classes matter at all. Note that the user is specifying the runtime class explicitly in the pod. Hence, we choose SCCs that contain that in `RequiredRuntimeClasses`. Why should the other runtime classes in the set influence the choice?

I would go this way:

- `points([..., "*", ...]) = 4`
- otherwise: `points(X) = max(len(X), 3)`

It orders SCCs with up to 3 items by length. I don't think it is of value to consider bigger sets. That will never happen in practice.

Note that we also have to think about precedence of these points to the existing ones. If you look into https://github.com/openshift/apiserver-library-go/blob/master/pkg/securitycontextconstraints/util/sort/byrestrictions.go and compare with the other point values, where do you see runtime classes?

- if you have the choice between: a) privileged + `["foo"]` b) unprivileged, `["*"]` – for a pod that does not require privileged, but wants runtime class `"foo"`?
- if you have the choice between: a) hostPort + `["foo"]` b) no host port, `["*"]`  – for a pod that does not request host ports, but wants runtime class `"foo"`?
- if you have the choice between: a) hostNetwork + `["foo"]` b) pod network, `["*"]`  – for a pod that does not require host networking, but wants runtime class `"foo"`?
- if you have the choice between: a) hostPath mount + `["foo"]` b) no hostPath mounts, `["*"]`  – for a pod that does not request a hostPath, but wants runtime class `"foo"`?
- if you have the choice between: a) runAsAnyUserPoints + `["foo"]` b) runAsUserPoints, `["*"]`  – for a pod that does not specify the user, but wants runtime class `"foo"`?
- if you have the choice between: a) all capabilities + `["foo"]` b) no extra capabilities, `["*"]`  – for a pod that does not ask for additional capabilities, but wants runtime class `"foo"`?

Reply:
I think I could use some help here. I have added some open questions about how to weight them relatively when we don't know each field. I am still not convinced we need some Enum type, and to handle apiserver drift, to be able to accurately describe the relative weight among each of the runtime classes.

I am also pretty sure running a container in a VM, or in a user namespace is more secure than dropping all capabilities, regardless of what that capability is (see https://man7.org/linux/man-pages/man7/user_namespaces.7.html). How do we handle the fact that [capMinPoints](https://github.com/openshift/apiserver-library-go/blob/master/pkg/securitycontextconstraints/util/sort/byrestrictions.go#L53) is 0?

Reply:
> How do we handle the fact that capMinPoints is 0?

Why do we have to? We might have to shift all cap value up with some offset if you want to put runtime classes "before".

> I am also pretty sure running a container in a VM, or in a user namespace is more secure than dropping all capabilities, regardless of what that capability is

So everything other than default `""` is more secure, independent from capabilities? This suggests we have to give non-default runtime classes more than `capMaxPoints = 9999`.

What about the other 5 questions in my comment above?

Reply:
About VM+any capability is more secure than non-VM+no-capability, to be very clear (and as a corollay: "we need enums"):

I don't think this order matters, and hence the corollary doesn't matter either: every pod either has empty runtime class (= default) or it has an explicit runtime class string specified. If that is true, every admission request will either
1. consider all SCCs **with default** class (among the required ones), 
2. or it will consider all SCCs **with some other**  runtime class. 
In either case, the other listed runtime class in those considered SCCs should not influence the order (because they don't matter for the specified runtime class).

Or in other words: the runtime class is not chosen by admission, but it only filters the set of applicable SCCs.

Reply:
> > How do we handle the fact that capMinPoints is 0?
> 
> Why do we have to? We might have to shift all cap value up with some offset if you want to put runtime classes "before".
> 

Ah I didn't know we were so flexible with moving the other options points arounds, sounds good.

> > I am also pretty sure running a container in a VM, or in a user namespace is more secure than dropping all capabilities, regardless of what that capability is
> 
> So everything other than default `""` is more secure, independent from capabilities? This suggests we have to give non-default runtime classes more than `capMaxPoints = 9999`.
> 
I think it's closer that: running a container in a vm or user namespace makes capMaxPoints for that pod really equal 0. Like, it takes away all of the security risk of capabilities. Same goes for runAsAnyUser. These runtime classes nullify the risk. 

> What about the other 5 questions in my comment above?

The other options (privileged, hostPort, hostPath, hostNetwork) + either VM or user ns is riskier than without those security options and "". Neither do anything to mitigate these security capabilities.


My general issue with this approach of ascribing value to runtime classes in general is that we are basically hard coding a value for a very flexible field. 

For instance, runtime classes could be the way we introduce [crun](github.com/containers/crun) into openshift, allowing users to opt into using crun when it becomes a tech preview. Technically, the runtime class that would use crun is no safer than runc (they are basically a 1:1 mapping). 

Do you have insight on how to future proof the field?